afqinsight/nn/performance_utils.py

```diff
@@ -0,0 +1,328 @@
+    train_dataset = train_dataset.batch(batch_size)
+    val_dataset = val_dataset.batch(batch_size)
+
+    return train_dataset, X_test, X_train, y_test, val_dataset
+
+
+def test_tensorflow_model(
+    model, train_dataset, val_dataset, X_test, y_test, n_epochs=100
+):
+    lr = 0.0001
+
+    print("XXXXXXXX")
+    model.compile(
+        loss="mean_squared_error",
+        optimizer=tf.keras.optimizers.Adam(learning_rate=lr),
+        metrics=[
+            "mean_squared_error",
+            tf.keras.metrics.RootMeanSquaredError(name="rmse"),
+            "mean_absolute_error",
+        ],
+    )
+
+    ckpt_filepath = tempfile.NamedTemporaryFile(suffix=".weights.h5").name
+    ckpt = tf.keras.callbacks.ModelCheckpoint(
+        filepath=ckpt_filepath,
+        monitor="val_loss",
+        verbose=0,
+        save_best_only=True,
+        save_weights_only=True,
+        mode="auto",
+    )
+
+    early_stopping = tf.keras.callbacks.EarlyStopping(
+        monitor="val_loss", min_delta=0.001, mode="min", patience=100
+    )
+
+    reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(
+        monitor="val_loss", factor=0.5, patience=20, verbose=1
+    )
+
+    callbacks = [early_stopping, ckpt, reduce_lr]
+
+    history = model.fit(
+        train_dataset,
+        epochs=n_epochs,
+        validation_data=val_dataset,
+        callbacks=callbacks,
+        verbose=2,
+    )
+
+    assert history.history[
+        "loss"
+    ], "Training history is empty. Model might not have run."
+    assert not np.isnan(history.history["loss"][-1]), "Final training loss is NaN."
+    assert not np.isinf(history.history["loss"][-1]), "Final training loss is Inf."
+
+    results = model.evaluate(X_test.astype(np.float32), y_test.astype(np.float32))
+
+    assert not np.isnan(results[0]), "Test loss is NaN."
+    assert not np.isinf(results[0]), "Test loss is Inf."
+
+    print(
+        f"Test Results - Loss: {results[0]}, RMSE: {results[1]}, MAE: {results[2]},"
+        f"MSE: {results[3]}"
+    )
+
+
+def get_pytorch_dataset():
+    dataset = AFQDataset.from_study("hbn")
+    dataset.drop_target_na()
+    imputer = dataset.model_fit(SimpleImputer(strategy="median"))
+    dataset = dataset.model_transform(imputer)
+    torch_dataset = dataset.as_torch_dataset(
+        bundles_as_channels=True, channels_last=False
+    )
+    train_dataset, test_dataset = torch.utils.data.random_split(
+        torch_dataset,
+        [
+            int(0.8 * len(torch_dataset)),
+            len(torch_dataset) - int(0.8 * len(torch_dataset)),
+        ],
+    )
+    train_dataset, val_dataset = torch.utils.data.random_split(
+        train_dataset,
+        [
+            int(0.8 * len(train_dataset)),
+            len(train_dataset) - int(0.8 * len(train_dataset)),
+        ],
+    )
+    imputer = train_dataset.model_fit(SimpleImputer(strategy="median"))
+
+    return train_dataset, val_dataset, test_dataset
+
+
+def get_dataloder_from_dataset(dataset, batch_size=32):
```

Comment:
Needs a docstring, and in particular what is "dataset"? I believe it's a pytorch dataset?